resolwe/flow/tests/test_purge.py

```diff
@@ -474,3 +474,47 @@ def test_remove_same_location(self):
             purge.purge_all(delete=True)
             rmtree_mock.assert_called_once_with(
                 os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], location))
+
+    # This patch is required so that the manager is not invoked while saving Data.
+    @disable_auto_calls()
+    def test_remove_storage(self):
+        data_1 = Data.objects.create(**self.data)
+        DataLocation.objects.create(location=str(data_1.pk)).data.add(data_1)
+        # data_1_location = data_1.location.location
+        data_1.status = Data.STATUS_DONE
+        data_1.output = {'sample': {'file': 'test-file'}}
+        self.create_test_file(data_1.location, 'test-file')
+        data_1.save()
+
+        storage_1 = data_1.storages.create(contributor=self.user, json={})
+        data_1.storages.create(contributor=self.user, json={})
+
+        data_2 = Data.objects.create(**self.data)
+        DataLocation.objects.create(location=str(data_2.pk)).data.add(data_2)
+        # data_2_location = data_2.location.location
+        data_2.status = Data.STATUS_DONE
+        data_2.output = {'sample': {'file': 'test-file'}}
+        self.create_test_file(data_2.location, 'test-file')
+        data_2.storages.add(storage_1)
+        data_2.save()
+
+        self.assertEqual(Data.objects.count(), 2)
+        self.assertEqual(Storage.objects.count(), 2)
+
+        # Delete first object.
+        data_1.delete()
```

Comment:
I was thinking about something like:

```
def delete(...):
    storages = list(self.storage.all())
    super().delete(...)
    for storage in storages:
        <delete if orphaned>
```

Reply:
Doesn't this have a significant performance impact? Could this produce a timeout?

Reply:
Does it even make sense to have storage purge then?

Reply:
>Does it even make sense to have storage purge then?

Oh, it does because there's other ways for data object to be deleted (instead via `Data.delete()`).